packages/shared/blockchain/evm-based/src/lib/services/evm-based.custodial.ts

```diff
@@ -72,60 +81,144 @@ export const evmBasedCustodial = () => {
+        fromPrivateKey: body.fromPrivateKey,
+        signatureId: body.signatureId,
+        index: body.index,
+        contractAddress: body.custodialAddress,
+        methodName: 'transfer',
+        params: [
+          body.tokenAddress || '0x000000000000000000000000000000000000dEaD',
+          body.contractType,
+          body.recipient,
+          `0x${amount.toString(16)}`,
+          `0x${new BigNumber(tokenId).toString(16)}`,
+        ],
+        methodABI: CustodialFullTokenWallet.abi.find((a) => a.name === 'transfer'),
+      }
+
       // @TODO OPENAPI NATIVE_ASSET
-      if (body.contractType === TransferCustodialWallet.contractType._3) {
+      if (body.contractType === NATIVE_ASSET) {
         amount = amount.multipliedBy(new BigNumber(10).pow(decimals))
         // @TODO OPENAPI FUNGIBLE_TOKEN
-      } else if (body.contractType === TransferCustodialWallet.contractType._0) {
+      } else if (body.contractType === FUNGIBLE_TOKEN) {
         tokenId = new BigNumber(0)
         amount = amount.multipliedBy(
-          new BigNumber(10).pow(await getContractDecimals(body.tokenAddress, provider, testnet)),
+          new BigNumber(10).pow(await getContractDecimals(body.tokenAddress!, web3, provider, testnet)),
         )
       }
-      r.params = [
-        body.tokenAddress || '0x000000000000000000000000000000000000dEaD',
-        body.contractType,
-        body.recipient,
-        `0x${amount.toString(16)}`,
-        `0x${new BigNumber(tokenId).toString(16)}`,
-      ]
-      r.methodABI = CustodialFullTokenWallet.abi.find((a) => a.name === 'transfer')
+
+      return await prepareSmartContractWriteMethodInvocation(
+        (body.chain === 'CELO' ? { ...r, feeCurrency: body.feeCurrency || Currency.CELO } : r) as SCBody,
+        web3,
+        provider,
+      )
+    },
+
+    prepareBatchTransferFromCustodialWalletAbstract: async <SCBody extends CallSmartContractMethod>(
+      body: ChainBatchTransferCustodialWallet,
+      web3: EvmBasedWeb3,
+      getContractDecimals: (
+        contractAddress: string,
+        web3: EvmBasedWeb3,
+        provider?: string,
+        testnet?: boolean,
+      ) => Promise<any>,
+      prepareSmartContractWriteMethodInvocation: (
+        r: SCBody,
+        web3: EvmBasedWeb3,
+        provider?: string,
+      ) => Promise<string>,
+      decimals: number,
+      testnet?: boolean,
+      provider?: string,
+    ) => {
+      const amounts = []
+      const tokenIds = []
+
+      for (let i = 0; i < body.contractType.length; i++) {
+        let amount = new BigNumber(body.amount ? body.amount[i] : 0)
+        let tokenId = new BigNumber(body.tokenId ? body.tokenId[i] : 0)
+        if (body.contractType[i] === NATIVE_ASSET) {
+          amount = amount.multipliedBy(new BigNumber(10).pow(decimals))
+        } else if (body.contractType[i] === NON_FUNGIBLE_TOKEN) {
+          amount = new BigNumber(0)
+        } else if (body.contractType[i] === FUNGIBLE_TOKEN && body.tokenAddress) {
```

Comment:
I dont see semi fungible tokens here

Reply:
I took this from the old sdk and it wasn't there either, I will try to implement it